keps/238-controller-revision/README.md

```diff
@@ -0,0 +1,394 @@
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+The following LWS API fields need to be added to support controller revision.
+
+```golang
+// LeaderWorkerSetStatus defines the observed state of LeaderWorkerSet
+type LeaderWorkerSetStatus struct {
+	// currentRevision, if not empty, indicates the version of lws
+	// used to generate the worker pods in sequence [0,currentReplicas)
+	CurrentRevision string `json:"currentRevision,omitempty"`
+
+	// updateRevision, if not empty, indicates the version of lws
+	// used to generate the worker pods in sequence 
+  // [replicas-updatedReplicas,replicas)
+	UpdateRevision string `json:"updateRevision,omitempty"`
+
+	// collisionCount is the count of hash collisions for the controller 
+	// revision uses this field as a collision avoidance mechanism 
+	// when it needs to create the name for the newest ControllerRevision.   
+	// +optional
+	CollisionCount *int32 `json:"collisionCount,omitempty"`
+}
+```
+
+### LWS controller
+
+The status of the revisions will be updated before rolling update starts.
+
+```golang
+func Reconcile() {
+	currentRevision, updateRevision, collisionCount, err := GetLeaderWorkerSetRevisions(ctx, r.Client, lws)
+	if err != nil {
+		log.Error(err, "Getting StatefulSet revisions")
+		return ctrl.Result{}, err
+	}
+	lws.Status.CurrentRevision = currentRevision.Name
+	lws.Status.UpdateRevision = updateRevision.Name
+	lws.Status.CollisionCount = new(int32)
+	lws.Status.CollisionCount = &collisionCount
+
+	partition, replicas, err := r.rollingUpdateParameters(ctx, lws)
+}
+```
+
+Once the update has been determined to be done, `currentRevision` will be set to be the value of `updateRevision`
```

Comment:
can you clarify when updateRevision gets created/set?

Reply:
`updateRevision` gets created when there isn't an equal revisions. So the pseudocode is

```golang
 func GetLeaderWorkerSetRevisions(ctx, r.Client, lws) {
    revisions := ListControllerRevisions()
    updateRevision := NewRevision(lws)
    equalRevisions := FindEqualRevisions(revisions, updateRevision)
    if len(equalRevisions) == 0 {
        CreateControllerRevision(lws, updateRevision)
    }
}
```

CurrentRevision doesn't get updated/changed unless `lws.CurrentRevision` is nil. So during an update, `updateRevision` and `currentRevision` will be different until the update is complete.